### ISSUE TITLE
[webview_flutter] Navigation delegate does not have the same behavior on Android and iOS

### ISSUE THREAD

**Issue by stevemoreau** [flutter/flutter#47218]
## TL;DR:

The navigation delegate is called several different times on iOS versus Android, which may make custom navigation handler difficult to maintain. The solution would be to rely on `isForMainFrame` to know if we enter the delegate on a real user interaction or not, but it may be confusing sometimes.

## Steps to Reproduce

1. Create a simple application with a webview with URL 'https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html' or clone and build the following project branch
```
git clone -b issue/navigation-delegate-ios-versus-android git@github.com:stevemoreau/flutter_webview.git
```
2. Run project on Android, click on the first link, and you should logs
```
I/flutter ( 3838): Entering build()
I/flutter ( 3838): Page started loading: https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html
I/flutter ( 3838): Page finished loading: https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html

// LINK CLICK HAPPENED HERE

I/flutter ( 3838): Entering navigationDelegate: requestUrl=https://www.mobizel.com/dummy-link-internal isForMainFrame=true
I/flutter ( 3838): Page started loading: https://www.mobizel.com/dummy-link-internal
I/flutter ( 3838): Page finished loading: https://www.mobizel.com/dummy-link-internal
```
3. Then, run project on iOS, click on the first link, you should see logs
```
flutter: Entering build()
flutter: Entering build()
flutter: Entering navigationDelegate: requestUrl=https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html isForMainFrame=true
flutter: Page started loading: https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html
flutter: Entering navigationDelegate: requestUrl=https://www.youtube.com/embed/p9rFDcS4F00 isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Page finished loading: https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html

// LINK CLICK HAPPENED HERE

flutter: Entering navigationDelegate: requestUrl=https://www.mobizel.com/dummy-link-internal isForMainFrame=true
flutter: Page started loading: https://www.mobizel.com/dummy-link-internal
flutter: Entering navigationDelegate: requestUrl=https://www.google.com/recaptcha/api2/anchor?ar=1&k=6LdO8pAUAAAAAK6Ei_JJmKQIztUcBZaBod-3NdCL&co=aHR0cHM6Ly93d3cubW9iaXplbC5jb206NDQz&hl=en&v=mhgGrlTs_PbFQOW4ejlxlxZn&size=invisible&cb=4xw6wzque43i isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Page finished loading: https://www.mobizel.com/dummy-link-internal
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
flutter: Entering navigationDelegate: requestUrl=about:blank isForMainFrame=false
```

**Target Platform:** iOS/Android
**Target OS version: Any**
**Devices: Any**
**Flutter configuration:** webview_flutter: ^0.3.18+1

## Logs

$ flutter run --verbose
[webview_navigation_delegate_android.txt](https://github.com/flutter/flutter/files/3973980/webview_navigation_delegate_android.txt)
[webview_navigation_delegate_ios.txt](https://github.com/flutter/flutter/files/3973981/webview_navigation_delegate_ios.txt)

```
$ flutter analyze
Analyzing flutter_webview...                          
No issues found! (ran in 9.5s)
```

```
$ flutter doctor -v
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.1 19B88, locale fr-FR)
    • Flutter version 1.12.13+hotfix.5 at /Users/steve/sdk/flutter
    • Framework revision 27321ebbad (6 days ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/steve/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (2 available)
    • Android SDK built for x86 • emulator-5554                            • android-x86 • Android 7.0 (API 24) (emulator)
    • iPhone                    • 5c21e6117ed37334bb86a3667172e7317c1f609e • ios         • iOS 13.2.3

• No issues found!
```

**Comment by mathieusachet** [comment]
same problem here

**Comment by bkraszewski** [comment]
Faced same issue, any update?

**Comment by claireliu14** [comment]
navigationDelegate 
on Android:  get changed url from NavigationRequest
on iOS: get about:blank

**Comment by john7ric** [comment]
web-view NavigationDelegate is a hit and miss for me . Not getting called in some scenarios.

**Comment by danagbemava-nc** [comment]
Reproducible on `master 2.5.0-6.0.pre.46` and `stable 2.2.3` with `webview_flutter: ^2.0.12`. 

<details>
<summary>code sample</summary>

```dart
// ignore_for_file: public_member_api_docs

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() => runApp(MaterialApp(home: WebViewExample()));

class WebViewExample extends StatefulWidget {
  @override
  _WebViewExampleState createState() => _WebViewExampleState();
}

class _WebViewExampleState extends State<WebViewExample> {
  final Completer<WebViewController> _controller =
      Completer<WebViewController>();

  @override
  Widget build(BuildContext context) {
    print('Entering build()');

    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter WebView example'),
      ),
      body: Builder(builder: (BuildContext context) {
        return WebView(
            initialUrl:
                'https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html',
            javascriptMode: JavascriptMode.unrestricted,
            onWebViewCreated: (WebViewController webViewController) {
              _controller.complete(webViewController);
            },
            gestureNavigationEnabled: true,
            javascriptChannels: <JavascriptChannel>[
              _toasterJavascriptChannel(context),
            ].toSet(),
            navigationDelegate: (NavigationRequest request) {
              print(
                  "\n+=======================================================================================+");
              print(
                '[${Platform.operatingSystem}] [navigationDelegate]: requestUrl=${request.url} isForMainFrame=${request.isForMainFrame}\n',
              );
              print(
                  "+=======================================================================================+\n");

              return NavigationDecision.navigate;
            },
            onPageStarted: (String url) {
              print(
                  "\n+=======================================================================================+");
              print('[${Platform.operatingSystem}] [onPageStarted] $url');
              print(
                  "+=======================================================================================+\n");
            },
            onPageFinished: (String url) {
              print(
                  "\n+=======================================================================================+");
              print('[${Platform.operatingSystem}] [onPageFinished] $url');
              print(
                  "+=======================================================================================+\n");
            });
      }),
      floatingActionButton: favoriteButton(),
    );
  }

  JavascriptChannel _toasterJavascriptChannel(BuildContext context) {
    return JavascriptChannel(
        name: 'Toaster',
        onMessageReceived: (JavascriptMessage message) {
          Scaffold.of(context).showSnackBar(
            SnackBar(content: Text(message.message)),
          );
        });
  }

  Widget favoriteButton() {
    return FutureBuilder<WebViewController>(
        future: _controller.future,
        builder: (BuildContext context,
            AsyncSnapshot<WebViewController> controller) {
          if (controller.hasData) {
            return FloatingActionButton(
              onPressed: () async {
                final String? url = await controller.data!.currentUrl();
                Scaffold.of(context).showSnackBar(
                  SnackBar(content: Text('Favorited $url')),
                );
              },
              child: const Icon(Icons.favorite),
            );
          }
          return Container();
        });
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.5.0-6.0.pre.46, on macOS 11.5.1 20G80 darwin-arm,
    locale en-GH)
    • Flutter version 2.5.0-6.0.pre.46 at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 41ff30c871 (16 minutes ago), 2021-08-10 23:03:48 -0700
    • Engine revision 4fef55db10
    • Dart version 2.14.0 (build 2.14.0-385.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554                        •
      android-arm64  • Android 12 (API 31) (emulator)
    • iPhone 12 (mobile)          • F2592D43-5FE3-46BB-8A2B-A38BA468DCE9 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)                • chrome                               •
      web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.5.1 20G80 darwin-arm, locale
    en-GH)
    • Flutter version 2.2.3 at /Users/nexus/dev/sdks/flutter
    • Framework revision f4abaa0735 (6 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554                        •
      android-arm64  • Android 12 (API 31) (emulator)
    • iPhone 12 (mobile)          • F2592D43-5FE3-46BB-8A2B-A38BA468DCE9 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)                • chrome                               •
      web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```
</details>

<details>
<summary>logs (removed verbose messages for android) </summary>

```
I/flutter (17692): Entering build()
I/flutter (17692): +=======================================================================================+
I/flutter (17692): [android] [onPageStarted] https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html
I/flutter (17692): +=======================================================================================+
I/chromium(17692): [INFO:CONSOLE(356)] "The deviceorientation events are blocked by permissions policy. See https://github.com/w3c/webappsec-permissions-policy/blob/master/features.md#sensor-features", source: https://www.youtube.com/s/player/4224c673/player_ias.vflset/en_US/base.js (356)
2
E/chromium(17692): [ERROR:web_contents_delegate.cc(225)] WebContentsDelegate::CheckMediaAccessPermission: Not supported.
W/cr_media(17692): Requires MODIFY_AUDIO_SETTINGS and RECORD_AUDIO. No audio device will be available for recording
I/flutter (17692): +=======================================================================================+
I/flutter (17692): [android] [onPageFinished] https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html
I/flutter (17692): +=======================================================================================+
I/flutter (17692): +=======================================================================================+
I/flutter (17692): [android] [navigationDelegate]: requestUrl=https://www.mobizel.com/dummy-link-internal isForMainFrame=true
I/flutter (17692): +=======================================================================================+
I/flutter (17692): +=======================================================================================+
I/flutter (17692): [android] [onPageStarted] https://www.mobizel.com/dummy-link-internal
I/flutter (17692): +=======================================================================================+
I/chromium(17692): [INFO:CONSOLE(5)] "JQMIGRATE: Migrate is installed, version 3.3.2", source: https://www.mobizel.com/wp-content/cache/autoptimize/js/autoptimize_140f12d0d82fccf2444f4a1942be7013.js (5)
I/flutter (17692): +=======================================================================================+
I/flutter (17692): [android] [onPageFinished] https://www.mobizel.com/dummy-link-internal
I/flutter (17692): +=======================================================================================+
I/.example.aug_e(17692): Background young concurrent copying GC freed 17795(785KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 3238KB/4093KB, paused 92us total 107.835ms
```

```
flutter: Entering build()
flutter:
+=======================================================================================+
flutter: [ios] [navigationDelegate]: requestUrl=https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html isForMainFrame=true
flutter: +=======================================================================================+
flutter:
+=======================================================================================+
flutter: [ios] [onPageStarted] https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html
flutter: +=======================================================================================+
flutter:
+=======================================================================================+
flutter: +=======================================================================================+
flutter:
+=======================================================================================+
flutter: [ios] [onPageFinished] https://app-mobile-webview.mobizel.com/webview_flutter/html_with_links.html
flutter: +=======================================================================================+
flutter:
+=======================================================================================+
flutter: [ios] [navigationDelegate]: requestUrl=about:blank isForMainFrame=false
flutter: +=======================================================================================+

```
</details>

**Comment by josefkvist** [comment]
What's the status here? I can see a commit, but no further updates. Can I help in any way?

**Comment by jdbenitezh** [comment]
the origin of the diferent behaviour is WKwebKit.  It can be solved manualy accessing to WKNavigationAction class of ios api. But flutter does not exposes this class.   Xamarin does it.

**Comment by scharfenort** [comment]
> the origin of the diferent behaviour is WKwebKit. It can be solved manualy accessing to WKNavigationAction class of ios api. But flutter does not exposes this class. Xamarin does it.

What does it mean? Do you have an workaround for us? What can we do? I am facing the same issue :-(

**Comment by sed1ka** [comment]
> I have solved my issue by changing a single line in NavigationDelegate. Actually, it was causing an error in iOS White Screen.
> 
> ```
>  ..setNavigationDelegate(
>         NavigationDelegate(
>           onProgress: (int progress) {
>             // Update loading bar.
>           },
>           onPageStarted: (String url) {},
>           onPageFinished: (String url) {},
>           onWebResourceError: (WebResourceError error) {},
>           onNavigationRequest: (NavigationRequest request) async {
>             if (request.url.startsWith('https://www.somesite.com/')) {
>               // do something...
>             } else if(request.url.startsWith('http://someothersite.com/')){
>               // prevent or do something else if this matches
>             }
>             return NavigationDecision.navigate;  // <-- Set it to navigate
>           },
>         ),
>       ),
> ```

on my case, onNavigationRequest never fire  )-:

**Comment by OhTerryTorres** [comment]
Not only is `onNavigationRequest` not firing for me, but `onPageFinished` is being called and `onPageStarted` is not.

**Comment by ftntming** [comment]
In my case, "onPageFinished" is not fired...

      ..setNavigationDelegate(
        NavigationDelegate(
          onPageFinished: (String url) async {
            debugPrint('Page finished loading: $url');
            currentUrl.setCurrentUrl(url);
          },
        ),
      )
